Framework/Core/include/Framework/ASoA.h

```diff
@@ -26,152 +27,362 @@
+#define O2HASH(_Str_)                              \
+  template <>                                      \
+  struct Hash<_Str_ ""_h> {                        \
+    static constexpr uint32_t hash = _Str_ ""_h;   \
+    static constexpr char const* const str{_Str_}; \
+  };
+
+/// Pre-declare Hash specialization for an origin string
+#define O2ORIGIN(_Str_)                                \
+  template <>                                          \
+  struct Hash<_Str_ ""_h> {                            \
+    static constexpr header::DataOrigin origin{_Str_}; \
+    static constexpr uint32_t hash = _Str_ ""_h;       \
+    static constexpr char const* const str{_Str_};     \
+  };
+
+/// Compile-time function to extract version from table signature string "DESC/#"
+static inline constexpr uint32_t version(const char* const str)
+{
+  if (str[0] == '\0') {
+    return 0;
+  }
+  size_t len = 0;
+  uint32_t res = 0;
+  while (str[len] != '/' && str[len] != '\0') {
+    ++len;
+  }
+  if (str[len - 1] == '\0') {
+    return -1;
+  }
+  for (auto i = len + 1; str[i] != '\0'; ++i) {
+    res = res * 10 + (int)(str[i] - '0');
+  }
+  return res;
+}
+
+/// Compile-time functions to extract description from table signature string "DESC/#"
+static inline constexpr std::string_view description_str(const char* const str)
+{
+  size_t len = 0;
+  while (len < 15 && str[len] != '/') {
+    ++len;
+  }
+  return std::string_view{str, len};
+}
 
+static inline constexpr header::DataDescription description(const char* const str)
+{
+  size_t len = 0;
+  while (len < 15 && str[len] != '/') {
+    ++len;
+  }
+  char out[16];
+  for (auto i = 0; i < 16; ++i) {
+    out[i] = 0;
+  }
+  std::memcpy(out, str, len);
+  return {out};
+}
+
+// Helpers to get strings from TableRef
+template <soa::TableRef R>
+consteval const char* label()
+{
+  return o2::aod::Hash<R.label_hash>::str;
+}
+
+template <soa::TableRef R>
+consteval const char* origin_str()
+{
+  return o2::aod::Hash<R.origin_hash>::str;
+}
+
+template <soa::TableRef R>
+consteval header::DataOrigin origin()
+{
+  return o2::aod::Hash<R.origin_hash>::origin;
+}
+
+template <soa::TableRef R>
+consteval const char* signature()
+{
+  return o2::aod::Hash<R.desc_hash>::str;
+}
+
+/// hash identification concepts
+template <typename T>
+concept is_aod_hash = requires(T t) { t.hash; t.str; };
+
+template <typename T>
+concept is_origin_hash = is_aod_hash<T> && requires(T t) { t.origin; };
+
+/// convert TableRef to a DPL source specification
+template <soa::TableRef R>
+static constexpr auto sourceSpec()
+{
+  return fmt::format("{}/{}/{}/{}", label<R>(), origin_str<R>(), description_str(signature<R>()), R.version);
+}
+} // namespace o2::aod
+
+namespace o2::soa
+{
+template <aod::is_aod_hash L, aod::is_aod_hash D, aod::is_origin_hash O, typename... Ts>
+class Table;
+
+/// Type-checking index column binding
 struct Binding {
   void const* ptr = nullptr;
   size_t hash = 0;
+  std::span<TableRef const> refs;
```

Comment:
Not for this PR, however why do you need a member for it? Can't you just have a function?

Reply:
When binding is created I can only store some aspects of the type, since binding itself is not templated. For now hash is enough, but I have some ideas to improve both setting and checking the index binding with this.